File: python/src/dmr_validation_framework/reports/upset_caller_support.R
#!/usr/bin/env Rscript
# Build ComplexUpset plots of DMR caller-support intersections.
#
# Input: a membership TSV with one row per consensus region, 0/1 columns per
# caller (the "sets"), and optional `region_id` and `context` columns.
# Output: upset_caller_support_overall.{pdf,png} and one plot per context.

arg_value <- function(name, default = NULL) {
  args <- commandArgs(trailingOnly = TRUE)
  key <- paste0("--", name)
  idx <- which(args == key)
  if (length(idx) == 0 || idx[length(idx)] >= length(args)) {
    return(default)
  }
  args[idx[length(idx)] + 1]
}

membership_path <- arg_value("membership")
out_dir <- arg_value("out-dir", ".")
colors_path <- arg_value("colors")
min_size <- as.integer(arg_value("min-size", "1"))
if (is.na(min_size) || min_size < 0) {
  min_size <- 1
}

# Optional caller -> hex color map (matches the Python report palette).
caller_colors <- list()
if (!is.null(colors_path) && file.exists(colors_path)) {
  color_table <- utils::read.delim(colors_path, stringsAsFactors = FALSE)
  if (all(c("caller", "color") %in% names(color_table))) {
    for (i in seq_len(nrow(color_table))) {
      caller_colors[[as.character(color_table$caller[i])]] <- as.character(color_table$color[i])
    }
  }
}

dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)
status_path <- file.path(out_dir, "upset_status.txt")
write_status <- function(text) {
  cat(text, "\n", file = status_path, append = FALSE)
}

if (is.null(membership_path) || !file.exists(membership_path)) {
  write_status(paste0("membership table not found: ", membership_path))
  quit(status = 3)
}

ok <- requireNamespace("ggplot2", quietly = TRUE) &&
  requireNamespace("ComplexUpset", quietly = TRUE)
if (!ok) {
  write_status("missing R packages: install ggplot2 and ComplexUpset")
  quit(status = 4)
}

suppressPackageStartupMessages({
  library(ggplot2)
  library(ComplexUpset)
})

df <- utils::read.delim(membership_path, check.names = FALSE, stringsAsFactors = FALSE)
meta_cols <- intersect(c("region_id", "context"), names(df))
set_cols_all <- setdiff(names(df), meta_cols)
if (length(set_cols_all) == 0) {
  write_status("no caller set columns found in membership table")
  quit(status = 5)
}
df[set_cols_all] <- lapply(df[set_cols_all], function(x) as.logical(as.integer(x)))

render_upset <- function(data, label) {
  set_cols <- set_cols_all[vapply(set_cols_all, function(col) any(data[[col]], na.rm = TRUE), logical(1))]
  if (length(set_cols) < 1 || nrow(data) == 0) {
    return(FALSE)
  }
  # One colored query per caller: colors the set-size bar and the matrix points
  # for that caller, using the shared report palette.
  queries <- list()
  for (col in set_cols) {
    if (!is.null(caller_colors[[col]])) {
      queries[[length(queries) + 1]] <- ComplexUpset::upset_query(
        set = col,
        fill = caller_colors[[col]],
        color = caller_colors[[col]]
      )
    }
  }
  # Render the intersection-size counts as vertical labels placed *above* every
  # bar (bar_number_threshold = Inf), with extra headroom on the y axis so the
  # numbers never overlap the bars or each other, even for the long tail of
  # small intersections.
  size_annotation <- ComplexUpset::intersection_size(
    bar_number_threshold = Inf,
    text = list(angle = 90, vjust = 0.5, hjust = -0.1, size = 2.6),
    text_colors = c(on_background = "black", on_bar = "black")
  ) + ggplot2::scale_y_continuous(expand = ggplot2::expansion(mult = c(0, 0.3)))
  plot <- ComplexUpset::upset(
    data,
    intersect = set_cols,
    name = "DMR caller support",
    min_size = min_size,
    sort_intersections_by = "cardinality",
    queries = queries,
    base_annotations = list("Intersection size" = size_annotation),
    set_sizes = ComplexUpset::upset_set_size()
  ) + ggplot2::labs(title = paste0("DMR caller support intersections: ", label))
  base <- file.path(out_dir, paste0("upset_caller_support_", label))
  # Widen the figure with the number of intersection bars actually shown so the
  # columns (and their vertical count labels) stay legible instead of crowding.
  codes <- apply(data[set_cols], 1, function(r) paste(as.integer(r), collapse = ""))
  n_intersections <- sum(table(codes) >= min_size)
  width <- max(7, min(26, n_intersections * 0.34), length(set_cols) * 1.6)
  tryCatch(
    {
      ggplot2::ggsave(paste0(base, ".pdf"), plot, width = width, height = 6, limitsize = FALSE)
      ggplot2::ggsave(paste0(base, ".png"), plot, width = width, height = 6, dpi = 200, limitsize = FALSE)
      TRUE
    },
    error = function(e) {
      cat("failed to render ", label, ": ", conditionMessage(e), "\n", file = status_path, append = TRUE)
      FALSE
    }
  )
}

built <- character()
if (render_upset(df, "overall")) {
  built <- c(built, "overall")
}
if ("context" %in% names(df)) {
  for (ctx in sort(unique(df$context))) {
    ctx_label <- gsub("[^A-Za-z0-9_]+", "_", as.character(ctx))
    if (render_upset(df[df$context == ctx, , drop = FALSE], ctx_label)) {
      built <- c(built, ctx_label)
    }
  }
}

write_status(paste0("built: ", paste(built, collapse = ", ")))
quit(status = 0)

File: python/src/dmr_validation_framework/reports/upset_caller_support.py
#!/usr/bin/env python
"""UpSet plots of DMR caller-support intersections (ComplexUpset, R).

Builds the caller-support membership table from a consensus support matrix (or,
when none exists yet, directly from the per-DMR caller rows the caller summary
writes) and renders ComplexUpset plots via a small R helper. The R step is
optional: if Rscript or the R packages are unavailable, the membership table is
still written and the step exits cleanly with a note.
"""

from __future__ import annotations

import argparse
import shutil
import subprocess
import sys
from pathlib import Path

import pandas as pd

from dmr_validation_framework.core.io import find_preferred_file, read_table

SUPPORT_MATRIX_NAMES = [
    "dmr_caller_support_matrix.tsv",
    "caller_support_matrix.tsv",
    "dmr_method_support_matrix.tsv",
    "external_caller_support_matrix.tsv",
]


def build_arg_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--out-dir", required=True, help="Directory for the membership table and UpSet figures.")
    parser.add_argument("--support-matrix", help="Consensus support matrix TSV (auto-discovered if omitted).")
    parser.add_argument("--caller-rows", help="caller_dmr_rows_for_plots.tsv, used to build a support matrix if none exists.")
    parser.add_argument("--rscript", help="Path to Rscript/Rscript.exe for the ComplexUpset step.")
    parser.add_argument("--overlap-threshold", type=float, default=0.5)
    parser.add_argument("--min-size", type=int, default=1, help="Minimum intersection size shown in the UpSet plot.")
    return parser


def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
    return build_arg_parser().parse_args(argv)


def _read(path: Path | None) -> pd.DataFrame:
    if path is None or not Path(path).exists():
        return pd.DataFrame()
    try:
        return read_table(path)
    except Exception:
        return pd.DataFrame()


def support_matrix_from_caller_rows(caller_rows: pd.DataFrame, overlap_threshold: float) -> pd.DataFrame:
    """Run the canonical consensus clustering on the flat per-DMR caller rows."""
    from bsx2.analysis.dmr_harmonization import build_caller_support_matrix

    df = caller_rows.copy()
    # _caller / _context are the authoritative per-row caller and context written
    # by the caller summary; prefer them over any canonical columns.
    if "_caller" in df.columns:
        df["source_caller"] = df["_caller"]
    elif "source_caller" not in df.columns and "caller" in df.columns:
        df["source_caller"] = df["caller"]
    if "_context" in df.columns:
        df["context"] = df["_context"]
    required = {"chrom", "start", "end", "context", "source_caller"}
    if not required.issubset(df.columns):
        return pd.DataFrame()
    return build_caller_support_matrix([df], overlap_threshold=overlap_threshold)


def membership_from_support(support: pd.DataFrame) -> pd.DataFrame:
    """One-hot caller membership per consensus region, preserving real names."""
    if support.empty or "supporting_callers" not in support.columns:
        return pd.DataFrame()
    rows: list[dict] = []
    callers: list[str] = []
    for value in support["supporting_callers"].dropna().astype(str):
        for caller in (item.strip() for item in value.split(",")):
            if caller and caller not in callers:
                callers.append(caller)
    if not callers:
        return pd.DataFrame()
    for idx, row in support.reset_index(drop=True).iterrows():
        present = {item.strip() for item in str(row.get("supporting_callers", "")).split(",") if item.strip()}
        record: dict = {"region_id": str(row.get("region_id", row.get("harmonized_region_id", idx)))}
        if "context" in support.columns:
            record["context"] = str(row.get("context", "NA"))
        for caller in callers:
            record[caller] = int(caller in present)
        rows.append(record)
    return pd.DataFrame(rows)


def resolve_rscript(path: str | None) -> str | None:
    if path:
        candidate = Path(path)
        if candidate.exists():
            return str(candidate)
        found = shutil.which(path)
        if found:
            return found
    return shutil.which("Rscript")


def run(args: argparse.Namespace) -> int:
    out_dir = Path(args.out_dir)
    out_dir.mkdir(parents=True, exist_ok=True)

    # Source precedence (deterministic): an explicit support matrix, then the
    # current caller rows (built fresh so the UpSet reflects exactly the callers
    # of this run), and only as a last resort an auto-discovered matrix. Auto-
    # discovery is for standalone CLI use; in a pipeline it can otherwise pick up
    # a stale/target matrix (e.g. one with an "internal" set), which silently
    # mislabels the plot.
    support = pd.DataFrame()
    if args.support_matrix:
        support = _read(Path(args.support_matrix))
    elif args.caller_rows and Path(args.caller_rows).exists():
        caller_rows = _read(Path(args.caller_rows))
        if not caller_rows.empty:
            support = support_matrix_from_caller_rows(caller_rows, args.overlap_threshold)
            if not support.empty:
                support.to_csv(out_dir / "dmr_caller_support_matrix.tsv", sep="\t", index=False)
    if support.empty:
        support = _read(find_preferred_file(SUPPORT_MATRIX_NAMES))

    membership = membership_from_support(support)
    if membership.empty:
        (out_dir / "upset_status.txt").write_text(
            "no caller-support membership could be built (no support matrix or caller rows).\n",
            encoding="utf-8",
        )
        print("upset: no membership table to plot", file=sys.stderr)
        return 0

    membership_path = out_dir / "upset_membership.tsv"
    membership.to_csv(membership_path, sep="\t", index=False)

    # Shared palette: one stable color per caller, matching the rest of the reports.
    from dmr_validation_framework.reports.palette import caller_color

    set_cols = [c for c in membership.columns if c not in {"region_id", "context"}]
    colors_path = out_dir / "upset_colors.tsv"
    pd.DataFrame(
        {"caller": set_cols, "color": [caller_color(c) for c in set_cols]}
    ).to_csv(colors_path, sep="\t", index=False)

    rscript = resolve_rscript(args.rscript)
    if rscript is None:
        (out_dir / "upset_status.txt").write_text(
            "Rscript not found; wrote membership table only. Install R + ComplexUpset to render UpSet plots.\n",
            encoding="utf-8",
        )
        print("upset: Rscript unavailable; wrote membership table only", file=sys.stderr)
        return 0

    r_script = Path(__file__).with_name("upset_caller_support.R")
    command = [
        rscript,
        str(r_script),
        "--membership",
        str(membership_path),
        "--out-dir",
        str(out_dir),
        "--colors",
        str(colors_path),
        "--min-size",
        str(args.min_size),
    ]
    completed = subprocess.run(command, check=False)
    if completed.returncode != 0:
        print(f"upset: ComplexUpset step exited with {completed.returncode}; see {out_dir / 'upset_status.txt'}", file=sys.stderr)
    else:
        print(f"upset figures: {out_dir}")
    return 0


def main(argv: list[str] | None = None) -> int:
    return run(parse_args(argv))


if __name__ == "__main__":
    raise SystemExit(main())
